hibernate-core/src/main/java/org/hibernate/annotations/TenantId.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.annotations;
+
+import org.hibernate.tuple.TenantIdGeneration;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Identifies a field of an entity that holds a tenant id
+ * in discriminator-based multitenancy.
+ *
+ * @author Gavin King
+ */
+@ValueGenerationType(generatedBy = TenantIdGeneration.class)
+@Target({TYPE, METHOD, FIELD})
```

Comment:
Do you think it makes sense to allow `@TenantId(columnName = "tenant_id")` on the class level or would we require the tenant id to always be mapped?